### ISSUE TITLE
Window Posts no longer work for sectors on a read-only filesystem

### ISSUE THREAD

**Issue by clinta** [filecoin-project/rust-fil-proofs#1629]
Something changed with v12 which seems to require writing to the sector store for window posts.

See: https://github.com/filecoin-project/lotus/issues/9298

Bisecting lotus narrowed down the issue to the upgrade in the rust-fil-proofs from 11.1 to 12.0.

This can be tested using `lotus-bench`, with the sealed secctor and cache located on a read-only filesystem. In previous versions this works without issue. After https://github.com/filecoin-project/lotus/commit/fa04496c81a7348e4ca75b3c0b9299cab4e898ec it fails with the following:

```
2022-09-14T03:41:29.046 INFO filcrypto::util::types > generate_fallback_sector_challenges: start
2022-09-14T03:41:29.046 INFO filecoin_proofs::api::post_util > generate_sector_challenges:start
2022-09-14T03:41:29.046 INFO filecoin_proofs::api::post_util > generate_sector_challenges:finish
2022-09-14T03:41:29.046 INFO filcrypto::util::types > generate_fallback_sector_challenges: end
2022-09-14T03:41:29.046 INFO filcrypto::util::types > generate_single_vanilla_proof: start
2022-09-14T03:41:29.046 INFO filecoin_proofs::api::post_util > generate_single_vanilla_proof:start: SectorId(24456)
2022-09-14T03:41:29.047 TRACE filecoin_proofs::types::private_replica_info > post: base tree size 153391689, base tree leafs 134217728, rows_to_discard 2, arities [8, 8, 0]
2022-09-14T03:41:29.047 INFO filcrypto::util::types > generate_single_vanilla_proof: end
2022-09-14T03:41:29.047Z	WARN	lotus-bench	lotus-bench/main.go:114	generate_single_vanilla_proof: merkle_tree failed: SectorId(24456)

Caused by:
    Read-only file system (os error 30)
```

**Comment by solopine** [comment]
WinningPost can also fail if WinningPost machine uses readonly storage.

**Comment by cryptonemo** [comment]
This part of the proofs code did not change (particularly around this point, though we never design or test with read-only storage in any cases).  Check `dmesg` on your system and see if you can better locate the cause. A common one may be that the system is trying to read past the end of storage (depends on how/why the storage became read-only).  I suspect a reboot and disk check may help.

**Comment by cryptonemo** [comment]
> https://github.com/filecoin-project/rust-fil-proofs/pull/1624 seems like it could be a factor. Changing the library used to mmap the files, the new library may be trying to mount those files read/write.

Interesting, and possibly.  That said, we had never aimed specifically to support read-only systems, but may be able to resolve this particular difference.  cc: @vmx @shawnrader

**Comment by clinta** [comment]
I'd love to see some tests added for this. For a large storage provider with many storage servers, it's not unheard of that a store may need to be mounted read-only while troubleshooting is done and backups are restored.

Additionally a provider may chose to mount storage on wdpost workers read-only, just to reduce the chances of corruption by limiting the processes that can write to the store.

**Comment by solopine** [comment]
I hope window Post and winning Post should support read-only storage. 1. for safe access. 2. for disk IO performance.  In my lab, when doing sealing (writing data) and windowPost on the same storage disk at the same time, windowPost of v1.17.1  takes longer time then before. I suppose disk-write may lock more resources then disk-read.

**Comment by storojs72** [comment]
PR with possible workaround for this issue has been merged. Now we need to deliver it to lotus

**Comment by cryptonemo** [comment]
Resolved by #1630
